### PR TITLE
add MoE code review

### PR DESCRIPTION
**Pull request by long8v**

### CODE REVIEW

File: MoE/moe.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+
+        if x.shape[0] == 1:
+            return torch.Tensor([0]).to(self.device)
+        return x.float().var() / (x.float().mean()**2 + eps)
+
+
+    def _gates_to_load(self, gates):
+        """Compute the true load per expert, given the gates.
+        The load is the number of examples for which the corresponding gate is >0.
```
[the remark this answers is not in the document]
Comment (long8v) [reply MoE/moe.py:174]: https://arxiv.org/pdf/1701.06538.pdf

File: MoE/moe.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        extra_training_loss: a scalar.  This should be added into the overall
+        training loss of the model.  The backpropagation of this loss
+        encourages all experts to be approximately equally used across a batch.
+        """
+        gates, load = self.noisy_top_k_gating(x, self.training)
+        # calculate importance loss
+        importance = gates.sum(0)
+        #
+        loss = self.cv_squared(importance) + self.cv_squared(load)
+        loss *= loss_coef
```
Comment (long8v) [inline MoE/moe.py:272]: importance에 대한 loss와 load에 대한 loss 합침

File: MoE/moe.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        """Noisy top-k gating.
+          See paper: https://arxiv.org/abs/1701.06538.
+          Args:
+            x: input Tensor with shape [batch_size, input_size]
+            train: a boolean - we only add noise at training time.
+            noise_epsilon: a float
+          Returns:
+            gates: a Tensor with shape [batch_size, num_experts]
+            load: a Tensor with shape [num_experts]
+        """
```
Comment (long8v) [inline MoE/moe.py:228]: top-k gate noisy 하게 만드는 부분

File: MoE/moe.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            logits = clean_logits
+
+        # calculate topk + 1 that will be needed for the noisy gates
+        top_logits, top_indices = logits.topk(min(self.k + 1, self.num_experts), dim=1)
+        top_k_logits = top_logits[:, :self.k]
+        top_k_indices = top_indices[:, :self.k]
+        top_k_gates = self.softmax(top_k_logits)
+
+        zeros = torch.zeros_like(logits, requires_grad=True).to(self.device)
+        gates = zeros.scatter(1, top_k_indices, top_k_gates).to(self.device)
```
Comment (long8v) [inline MoE/moe.py:245]: https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_
